opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+                with tracer.span("two") as child:
+                    child.add_event("child's event")
+                    tracer.get_current_span()  # returns child
+                tracer.get_current_span()      # returns parent
+            tracer.get_current_span()          # returns the previously active span
+
+        Args:
+            name: The name of the span to be created.
+            parent: This span's parent.
+
+        Raises:
+            ValueError: if ``name`` is null.
+        """
+        raise NotImplementedError
+
+    def create_span(self, name: str, parent: Span=None) -> Span:
+        """Create a new span as a child of the currently active span.
+
+        If called with ``parent`` this method won't affect the tracer's
+        context.
+
+        See ``span`` for a context manager that controls the span's lifetime.
+
+        Args:
+            name: The name of the span to be created.
+            parent: This span's parent.
+
+        Raises:
+            ValueError: if ``name`` is null.
+        """
+        raise NotImplementedError
+
+
+class Span(object):
+    """A span represents a single operation within a trace.
+    """
+
+    def start(self) -> None:
+        """Set the current time as the span's start time.
+
+        Each span represents a single operation. The span's start time is the
+        wall time at which the operation started.
+
+        Only the first call to ``start`` should modify the span, and
+        implementations are free to ignore or raise on further calls.
+        """
+        raise NotImplementedError
+
+    def end(self) -> None:
+        """Set the current time as the span's end time.
+
+        The span's end time is the wall time at which the operation finished.
+
+        Only the first call to ``end`` should modify the span, and
+        implementations are free to ignore or raise on further calls.
+        """
+        raise NotImplementedError
+
+    def get_context(self) -> SpanContext:
+        """Get an immutable snapshot of this span.
+
+        Returns:
+            A :class:`.SpanContext` with a copy of this span's immutable state.
+        """
+        raise NotImplementedError
+
+
+class SpanContext(object):
+    """The state of a Span to propagate between processes.
+
+    This class includes the immutable attributes of a :class:`.Span` that must
+    be propagated to a span's children and across process boundaries.
+
+    Args:
+        trace_id: The ID of the trace that this span belongs to.
+        span_id: This span's ID.
+        options: Global trace options to propagate.
+        state: Global tracing-system-specific info to propagate.
+    """
+
+    def __init__(self,
+                 trace_id: str,
+                 span_id: str,
+                 options: TraceOptions,
+                 state: TraceState) -> SpanContext:
+        pass
+
+
+# TODO
+class TraceOptions(int):
+    pass
+
+
+# TODO
+class TraceState(dict):
```

Comment:
Like it! There were conversations whether we want `TraceState` or `Tracestate`, and at that time `Tracestate` won.
https://github.com/census-instrumentation/opencensus-java/pull/1300#discussion_r201773204

Each time people see `Tracestate`, they ask me (and I've been asked 20+ times) whether it is a typo. Personally I think we probably have made the wrong choice as a team and it would be great if we can fix it.

Reply:
It's called that in the W3C spec: https://www.w3.org/TR/trace-context/#tracestate-field (they have Tracestate and Traceparent). Though for consistency with `TraceOptions` and knowing that the `Tracestate` might not end up in the W3C serialization I would also prefer `TraceState`. But IMHO consistency across OpenTelemetry's languages is most important here, so if we keep `TraceState` we should also convice the Java team.